aws/resource_aws_appsync_graphql_api.go

```diff
@@ -0,0 +1,215 @@
+package aws
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appsync"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsAppsyncGraphqlApi() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsAppsyncGraphqlApiCreate,
+		Read:   resourceAwsAppsyncGraphqlApiRead,
+		Update: resourceAwsAppsyncGraphqlApiUpdate,
+		Delete: resourceAwsAppsyncGraphqlApiDelete,
+
+		Schema: map[string]*schema.Schema{
+			"authentication_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := strings.ToUpper(v.(string))
+					validTypes := []string{"API_KEY", "AWS_IAM", "AMAZON_COGNITO_USER_POOLS"}
+					for _, str := range validTypes {
+						if value == str {
+							return
+						}
+					}
+					errors = append(errors, fmt.Errorf("expected %s to be one of %v, got %s", k, validTypes, value))
+					return
```

Comment:
Would you mind using constants available in the SDK?
Also we have a built-in validation helper for this:

```go
ValidateFunc: validation.StringInSlice([]string{
  appsync.AuthenticationTypeApiKey,
  appsync.AuthenticationTypeAwsIam,
  appsync.AuthenticationTypeAmazonCognitoUserPools,
}, false),
```
😉